storage/types.go

```diff
@@ -277,6 +282,34 @@ func (n *NodeID) String() string {
 	return r.String()
 }
 
+// AsKey returns a string identifier for this NodeID suitable for
+// short term use e.g. as a Map key. It is more efficient to use this than
+// String() as it's not constrained to return a binary string.
+func (n *NodeID) AsKey() string {
```

Comment:
A string in Go can hold arbitrary bytes. Why not just copy `n.PrefixLenBits + n.Path` as bytes for more efficiency?

Reply:
Because this gives a short ID that's human readable.

Reply:
Why does it need to be human readable? Its primary usage is being a key for in-memory hashmaps, isn't it? The only places where it needs to be human readable is tests and logs, but there you could print them with "%x" which would result in the same string as currently (also in tests the "golden" values can be hex-decoded).

Reply:
It seems like a nice property and it's fairly cheap.

Reply:
I don't think it's cheap on the relative scale. Compare:

```golang
func (n *NodeID) AsKey() string {
	var b strings.Builder
	fullBytes := n.PrefixLenBits / 8
	bitsLeft := n.PrefixLenBits % 8

	lb := make([]byte, 2)
	binary.LittleEndian.PutUint16(lb, uint16(n.PrefixLenBits)) // Or uint32, or varint.
	b.Write(lb[:])
	b.Write(n.Path[:fullBytes])
	// If there's bits left over write them out.
	if bitsLeft > 0 {
		bits := n.Path[fullBytes] & leftmask[bitsLeft]
		b.WriteByte(bits)
	}

	return b.String()
}
```

```
Before:	BenchmarkAsKey-12    	 5000000	       286 ns/op
After:	BenchmarkAsKey-12    	20000000	       109 ns/op
```

There is already one human readable format in `String()`. As another option, I think it would make sense to replace `String()` with your optimized version (I think there is no code other than in-memory stuff that relies on this format currently), and refactor `AsKey()` to return bytes packed to a string.

Reply:
Very very small aside. Could:
b.Write(lb[:])
be written as:
b.Write(lb)?

Reply:
Oh, I actually meant it to be `var lb [2]byte`, and then it would imply`lb[:]`.

Reply:
It's hard to be certain that it's not used for something important. My IDE says it isn't but `String()` can be called implicitly by `fmt` methods, which it might not be able to pick up. I think a `String()` method should only return something readable so we're stuck with it.